ImageBufferBackend has three overlapping copy-like functions

kkinnunen-apple opened the pull request:
#### 329f7da0cc3778b668fd606736876062da438036
<pre>
ImageBufferBackend has three overlapping copy-like functions
<a href="https://bugs.webkit.org/show_bug.cgi?id=261731">https://bugs.webkit.org/show_bug.cgi?id=261731</a>
rdar://115718413

Reviewed by Matt Woodrow.

ImageBuffer and -Backend function copyNativeImage(BackingStoreCopy)
is actually two rather distinct functions: copy and a reference.
copyNativeImageForDrawing() is equivalent to
copyNativeImage(DontBackingStore) and certain custom logic needed for
the call chain.

Move the decision to do a copy for draw or a reference for draw into
top-level GraphicsContext. This information will be also used in future
when implementing more correct display list retaining of backing stores
as well as removing the image buffer related remote drawing commands.

Make a clear distinction between creating a copy and creating a
reference to the backing store by calling the functions with different
names. Use the function in more places that obtain a drawable
NativeImage out of an ImageBuffer.

Removes ImageBuffer::draw because it was a redudundant addition
to the call-stack:
GraphicsContext::drawImageBuffer
    ImageBuffer::draw
       Graphicscontext::drawNativeImageInternal

This is work towards making the NativeImage the only image primitive
being drawn from, where as currently also ImageBuffer is a
pseudo-primitive.

* Source/WebCore/platform/graphics/CachedSubimage.cpp:
(WebCore::CachedSubimage::draw):
* Source/WebCore/platform/graphics/GraphicsContext.cpp:
(WebCore::GraphicsContext::GraphicsContext):
(WebCore::GraphicsContext::nativeImageForDrawing):
(WebCore::GraphicsContext::drawImageBuffer):
* Source/WebCore/platform/graphics/GraphicsContext.h:
(WebCore::GraphicsContext::GraphicsContext):
(WebCore::GraphicsContext::deferred const):
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::copyImageBufferToNativeImage):
(WebCore::ImageBuffer::copyNativeImage const):
(WebCore::ImageBuffer::createNativeImageReference const):
(WebCore::ImageBuffer::copyNativeImageForDrawing const): Deleted.
(WebCore::ImageBuffer::draw): Deleted.
* Source/WebCore/platform/graphics/ImageBuffer.h:
* Source/WebCore/platform/graphics/ImageBufferBackend.cpp:
(WebCore::ImageBufferBackend::sinkIntoNativeImage):
(WebCore::ImageBufferBackend::copyNativeImageForDrawing): Deleted.
* Source/WebCore/platform/graphics/ImageBufferBackend.h:
(WebCore::ImageBufferBackend::finalizeDrawIntoContext): Deleted.
* Source/WebCore/platform/graphics/SourceImage.cpp:
(WebCore::SourceImage::nativeImage const):
* Source/WebCore/platform/graphics/cairo/CairoOperations.cpp:
(WebCore::Cairo::drawShadowLayerBuffer):
(WebCore::Cairo::drawShadowImage):
(WebCore::Cairo::fillShadowBuffer):
* Source/WebCore/platform/graphics/cairo/GraphicsContextCairo.cpp:
(WebCore::GraphicsContextCairo::clipToImageBuffer):
(WebCore::GraphicsContextCairo::drawFocusRing): Deleted.
(WebCore::GraphicsContextCairo::drawLinesForText): Deleted.
(WebCore::GraphicsContextCairo::drawDotsForDocumentMarker): Deleted.
(WebCore::GraphicsContextCairo::translate): Deleted.
(WebCore::GraphicsContextCairo::didUpdateState): Deleted.
(WebCore::GraphicsContextCairo::concatCTM): Deleted.
(WebCore::GraphicsContextCairo::setCTM): Deleted.
(WebCore::GraphicsContextCairo::beginTransparencyLayer): Deleted.
(WebCore::GraphicsContextCairo::endTransparencyLayer): Deleted.
(WebCore::GraphicsContextCairo::clearRect): Deleted.
(WebCore::GraphicsContextCairo::strokeRect): Deleted.
(WebCore::GraphicsContextCairo::setLineCap): Deleted.
(WebCore::GraphicsContextCairo::setLineDash): Deleted.
(WebCore::GraphicsContextCairo::setLineJoin): Deleted.
(WebCore::GraphicsContextCairo::setMiterLimit): Deleted.
(WebCore::GraphicsContextCairo::clipOut): Deleted.
(WebCore::GraphicsContextCairo::rotate): Deleted.
(WebCore::GraphicsContextCairo::scale): Deleted.
(WebCore::GraphicsContextCairo::fillRoundedRectImpl): Deleted.
(WebCore::GraphicsContextCairo::fillRectWithRoundedHole): Deleted.
(WebCore::GraphicsContextCairo::drawPattern): Deleted.
(WebCore::GraphicsContextCairo::renderingMode const): Deleted.
(WebCore::GraphicsContextCairo::cr const): Deleted.
(WebCore::GraphicsContextCairo::layers): Deleted.
(WebCore::GraphicsContextCairo::pushImageMask): Deleted.
* Source/WebCore/platform/graphics/cairo/GraphicsContextCairo.h:
* Source/WebCore/platform/graphics/cairo/ImageBufferCairoSurfaceBackend.cpp:
(WebCore::ImageBufferCairoSurfaceBackend::copyNativeImage):
(WebCore::ImageBufferCairoSurfaceBackend::createNativeImageReference):
(WebCore::ImageBufferCairoSurfaceBackend::cairoSurfaceCoerceToImage):
* Source/WebCore/platform/graphics/cairo/ImageBufferCairoSurfaceBackend.h:
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::isDeferredForCGContext):
(WebCore::GraphicsContextCG::GraphicsContextCG):
(WebCore::GraphicsContextCG::clipToImageBuffer):
* Source/WebCore/platform/graphics/cg/ImageBufferCGBitmapBackend.cpp:
(WebCore::ImageBufferCGBitmapBackend::copyNativeImage):
(WebCore::ImageBufferCGBitmapBackend::createNativeImageReference):
* Source/WebCore/platform/graphics/cg/ImageBufferCGBitmapBackend.h:
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
(WebCore::ImageBufferIOSurfaceBackend::copyNativeImage):
(WebCore::ImageBufferIOSurfaceBackend::createNativeImageReference):
(WebCore::ImageBufferIOSurfaceBackend::copyNativeImageForDrawing): Deleted.
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.h:
* Source/WebCore/platform/graphics/cg/NativeImageCG.cpp:
(WebCore::NativeImage::draw):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
(WebCore::DisplayList::Recorder::Recorder):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.h:
(WebCore::DisplayList::Recorder::Recorder):
* Source/WebKit/Shared/RemoteLayerTree/CGDisplayListImageBufferBackend.h:
* Source/WebKit/Shared/RemoteLayerTree/CGDisplayListImageBufferBackend.mm:
(WebKit::CGDisplayListImageBufferBackend::copyNativeImage):
(WebKit::CGDisplayListImageBufferBackend::createNativeImageReference):
* Source/WebKit/Shared/WebImage.cpp:
(WebKit::WebImage::copyNativeImage const):
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferShareableBitmapBackend.cpp:
(WebKit::ImageBufferShareableBitmapBackend::copyNativeImage):
(WebKit::ImageBufferShareableBitmapBackend::createNativeImageReference):
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferShareableBitmapBackend.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
(WebKit::RemoteDisplayListRecorderProxy::RemoteDisplayListRecorderProxy):
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxy::copyNativeImage const):
(WebKit::RemoteImageBufferProxy::createNativeImageReference const):
(WebKit::RemoteImageBufferProxy::copyNativeImageForDrawing const): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferRemoteIOSurfaceBackend.cpp:
(WebKit::ImageBufferRemoteIOSurfaceBackend::copyNativeImage):
(WebKit::ImageBufferRemoteIOSurfaceBackend::createNativeImageReference):
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferRemoteIOSurfaceBackend.h:
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferShareableMappedIOSurfaceBackend.cpp:
(WebKit::ImageBufferShareableMappedIOSurfaceBackend::copyNativeImage):
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferShareableMappedIOSurfaceBackend.h:
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferShareableMappedIOSurfaceBitmapBackend.cpp:
(WebKit::ImageBufferShareableMappedIOSurfaceBitmapBackend::copyNativeImage):
(WebKit::ImageBufferShareableMappedIOSurfaceBitmapBackend::createNativeImageReference):
(WebKit::ImageBufferShareableMappedIOSurfaceBitmapBackend::copyNativeImageForDrawing): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferShareableMappedIOSurfaceBitmapBackend.h:

Canonical link: <a href="https://commits.webkit.org/268243@main">https://commits.webkit.org/268243@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/aa9c72a0d3f4d12ae892d2d019e3485da8ec89ff

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19121 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19468 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/20076 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20987 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17868 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/19321 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22756 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19601 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/19594 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19343 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19425 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16630 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21868 "Built successfully") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16624 "7 flakes 5 failures") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17415 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/23790 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17672 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17589 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21740 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/18155 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/15404 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17260 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4560 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21614 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17985 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mattwoodrow reviewed (commented):
Do you think you could expand the header comments in ImageBuffer (and maybe add equivalents to ImageBufferBackend)? I'm still not 100% clear on what the different semantics are.

Does `copyNativeImage` create an immutable snapshot of the current state (but could be implemented as a copy-on-write reference to the pixel data), whereas `createNativeImageReference` creates a live reference to the pixel data that will reflect the state of the ImageBuffer at the time when it's used?

What about `sinkIntoNativeImage`? When do we use that?

kkinnunen-apple committed:
ImageBufferBackend has three overlapping copy-like functions
https://bugs.webkit.org/show_bug.cgi?id=261731
rdar://115718413

Reviewed by Matt Woodrow.

ImageBuffer and -Backend function copyNativeImage(BackingStoreCopy)
is actually two rather distinct functions: copy and a reference.
copyNativeImageForDrawing() is equivalent to
copyNativeImage(DontBackingStore) and certain custom logic needed for
the call chain.

Move the decision to do a copy for draw or a reference for draw into
top-level GraphicsContext. This information will be also used in future
when implementing more correct display list retaining of backing stores
as well as removing the image buffer related remote drawing commands.

Make a clear distinction between creating a copy and creating a
reference to the backing store by calling the functions with different
names. Use the function in more places that obtain a drawable
NativeImage out of an ImageBuffer.

Removes ImageBuffer::draw because it was a redudundant addition
to the call-stack:
GraphicsContext::drawImageBuffer
    ImageBuffer::draw
       Graphicscontext::drawNativeImageInternal

This is work towards making the NativeImage the only image primitive
being drawn from, where as currently also ImageBuffer is a
pseudo-primitive.

* Source/WebCore/platform/graphics/CachedSubimage.cpp:
(WebCore::CachedSubimage::draw):
* Source/WebCore/platform/graphics/GraphicsContext.cpp:
(WebCore::GraphicsContext::GraphicsContext):
(WebCore::GraphicsContext::nativeImageForDrawing):
(WebCore::GraphicsContext::drawImageBuffer):
* Source/WebCore/platform/graphics/GraphicsContext.h:
(WebCore::GraphicsContext::GraphicsContext):
(WebCore::GraphicsContext::deferred const):
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::copyImageBufferToNativeImage):
(WebCore::ImageBuffer::copyNativeImage const):
(WebCore::ImageBuffer::createNativeImageReference const):
(WebCore::ImageBuffer::copyNativeImageForDrawing const): Deleted.
(WebCore::ImageBuffer::draw): Deleted.
* Source/WebCore/platform/graphics/ImageBuffer.h:
* Source/WebCore/platform/graphics/ImageBufferBackend.cpp:
(WebCore::ImageBufferBackend::sinkIntoNativeImage):
(WebCore::ImageBufferBackend::copyNativeImageForDrawing): Deleted.
* Source/WebCore/platform/graphics/ImageBufferBackend.h:
(WebCore::ImageBufferBackend::finalizeDrawIntoContext): Deleted.
* Source/WebCore/platform/graphics/SourceImage.cpp:
(WebCore::SourceImage::nativeImage const):
* Source/WebCore/platform/graphics/cairo/CairoOperations.cpp:
(WebCore::Cairo::drawShadowLayerBuffer):
(WebCore::Cairo::drawShadowImage):
(WebCore::Cairo::fillShadowBuffer):
* Source/WebCore/platform/graphics/cairo/GraphicsContextCairo.cpp:
(WebCore::GraphicsContextCairo::clipToImageBuffer):
(WebCore::GraphicsContextCairo::drawFocusRing): Deleted.
(WebCore::GraphicsContextCairo::drawLinesForText): Deleted.
(WebCore::GraphicsContextCairo::drawDotsForDocumentMarker): Deleted.
(WebCore::GraphicsContextCairo::translate): Deleted.
(WebCore::GraphicsContextCairo::didUpdateState): Deleted.
(WebCore::GraphicsContextCairo::concatCTM): Deleted.
(WebCore::GraphicsContextCairo::setCTM): Deleted.
(WebCore::GraphicsContextCairo::beginTransparencyLayer): Deleted.
(WebCore::GraphicsContextCairo::endTransparencyLayer): Deleted.
(WebCore::GraphicsContextCairo::clearRect): Deleted.
(WebCore::GraphicsContextCairo::strokeRect): Deleted.
(WebCore::GraphicsContextCairo::setLineCap): Deleted.
(WebCore::GraphicsContextCairo::setLineDash): Deleted.
(WebCore::GraphicsContextCairo::setLineJoin): Deleted.
(WebCore::GraphicsContextCairo::setMiterLimit): Deleted.
(WebCore::GraphicsContextCairo::clipOut): Deleted.
(WebCore::GraphicsContextCairo::rotate): Deleted.
(WebCore::GraphicsContextCairo::scale): Deleted.
(WebCore::GraphicsContextCairo::fillRoundedRectImpl): Deleted.
(WebCore::GraphicsContextCairo::fillRectWithRoundedHole): Deleted.
(WebCore::GraphicsContextCairo::drawPattern): Deleted.
(WebCore::GraphicsContextCairo::renderingMode const): Deleted.
(WebCore::GraphicsContextCairo::cr const): Deleted.
(WebCore::GraphicsContextCairo::layers): Deleted.
(WebCore::GraphicsContextCairo::pushImageMask): Deleted.
* Source/WebCore/platform/graphics/cairo/GraphicsContextCairo.h:
* Source/WebCore/platform/graphics/cairo/ImageBufferCairoSurfaceBackend.cpp:
(WebCore::ImageBufferCairoSurfaceBackend::copyNativeImage):
(WebCore::ImageBufferCairoSurfaceBackend::createNativeImageReference):
(WebCore::ImageBufferCairoSurfaceBackend::cairoSurfaceCoerceToImage):
* Source/WebCore/platform/graphics/cairo/ImageBufferCairoSurfaceBackend.h:
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::isDeferredForCGContext):
(WebCore::GraphicsContextCG::GraphicsContextCG):
(WebCore::GraphicsContextCG::clipToImageBuffer):
* Source/WebCore/platform/graphics/cg/ImageBufferCGBitmapBackend.cpp:
(WebCore::ImageBufferCGBitmapBackend::copyNativeImage):
(WebCore::ImageBufferCGBitmapBackend::createNativeImageReference):
* Source/WebCore/platform/graphics/cg/ImageBufferCGBitmapBackend.h:
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
(WebCore::ImageBufferIOSurfaceBackend::copyNativeImage):
(WebCore::ImageBufferIOSurfaceBackend::createNativeImageReference):
(WebCore::ImageBufferIOSurfaceBackend::copyNativeImageForDrawing): Deleted.
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.h:
* Source/WebCore/platform/graphics/cg/NativeImageCG.cpp:
(WebCore::NativeImage::draw):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
(WebCore::DisplayList::Recorder::Recorder):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.h:
(WebCore::DisplayList::Recorder::Recorder):
* Source/WebKit/Shared/RemoteLayerTree/CGDisplayListImageBufferBackend.h:
* Source/WebKit/Shared/RemoteLayerTree/CGDisplayListImageBufferBackend.mm:
(WebKit::CGDisplayListImageBufferBackend::copyNativeImage):
(WebKit::CGDisplayListImageBufferBackend::createNativeImageReference):
* Source/WebKit/Shared/WebImage.cpp:
(WebKit::WebImage::copyNativeImage const):
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferShareableBitmapBackend.cpp:
(WebKit::ImageBufferShareableBitmapBackend::copyNativeImage):
(WebKit::ImageBufferShareableBitmapBackend::createNativeImageReference):
* Source/WebKit/WebProcess/GPU/graphics/ImageBufferShareableBitmapBackend.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
(WebKit::RemoteDisplayListRecorderProxy::RemoteDisplayListRecorderProxy):
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxy::copyNativeImage const):
(WebKit::RemoteImageBufferProxy::createNativeImageReference const):
(WebKit::RemoteImageBufferProxy::copyNativeImageForDrawing const): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferRemoteIOSurfaceBackend.cpp:
(WebKit::ImageBufferRemoteIOSurfaceBackend::copyNativeImage):
(WebKit::ImageBufferRemoteIOSurfaceBackend::createNativeImageReference):
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferRemoteIOSurfaceBackend.h:
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferShareableMappedIOSurfaceBackend.cpp:
(WebKit::ImageBufferShareableMappedIOSurfaceBackend::copyNativeImage):
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferShareableMappedIOSurfaceBackend.h:
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferShareableMappedIOSurfaceBitmapBackend.cpp:
(WebKit::ImageBufferShareableMappedIOSurfaceBitmapBackend::copyNativeImage):
(WebKit::ImageBufferShareableMappedIOSurfaceBitmapBackend::createNativeImageReference):
(WebKit::ImageBufferShareableMappedIOSurfaceBitmapBackend::copyNativeImageForDrawing): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferShareableMappedIOSurfaceBitmapBackend.h:

Canonical link: https://commits.webkit.org/268243@main
